src/shared/machine-id-setup.c

```diff
@@ -250,30 +297,43 @@ int machine_id_setup(const char *root, sd_id128_t machine_id, MachineIdSetupFlag
 }
 
 int machine_id_commit(const char *root) {
-        _cleanup_close_ int fd = -EBADF, initial_mntns_fd = -EBADF;
-        const char *etc_machine_id;
         sd_id128_t id;
         int r;
 
-        /* Before doing anything, sync everything to ensure any changes by first-boot units are persisted.
-         *
-         * First, explicitly sync the file systems we care about and check if it worked. */
-        FOREACH_STRING(sync_path, "/etc/", "/var/") {
-                r = syncfs_path(AT_FDCWD, sync_path);
-                if (r < 0)
-                        return log_error_errno(r, "Cannot sync %s: %m", sync_path);
-        }
+        if (empty_or_root(root)) {
+                /* Before doing anything, sync everything to ensure any changes by first-boot units are
+                 * persisted.
+                 *
+                 * First, explicitly sync the file systems we care about and check if it worked. */
+                FOREACH_STRING(sync_path, "/etc/", "/var/") {
+                        r = syncfs_path(AT_FDCWD, sync_path);
+                        if (r < 0)
+                                return log_error_errno(r, "Cannot sync %s: %m", sync_path);
+                }
 
-        /* Afterwards, sync() the rest too, but we can't check the return value for these. */
-        sync();
+                /* Afterwards, sync() the rest too, but we can't check the return value for these. */
+                sync();
+        }
 
         /* Replaces a tmpfs bind mount of /etc/machine-id by a proper file, atomically. For this, the umount is removed
          * in a mount namespace, a new file is created at the right place. Afterwards the mount is also removed in the
          * original mount namespace, thus revealing the file that was just created. */
 
-        etc_machine_id = prefix_roota(root, "/etc/machine-id");
+        _cleanup_close_ int etc_fd = -EBADF;
+        _cleanup_free_ char *etc = NULL;
+        r = chase("/etc/", root, CHASE_PREFIX_ROOT|CHASE_MUST_BE_DIRECTORY, &etc, &etc_fd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open /etc/: %m");
+
+        _cleanup_free_ char *etc_machine_id = path_join(etc, "machine-id");
+        if (!etc_machine_id)
+                return log_oom();
 
-        r = path_is_mount_point(etc_machine_id);
+        r = fd_is_mount_point(etc_fd, "machine-id", /* flags= */ 0);
+        if (r == -ELOOP) {
```

Comment:
Hmm? Why would fd_is_mount_point() return ELOOP? With the latest kernel symlinks can be mountpoints after all, and we should handle it?